include/bout/coordinates.hxx

```diff
@@ -80,36 +87,46 @@ public:
    * Must be a better way so that Coordinates doesn't depend on Datafile
    */
   void outputVars(Datafile &file);
-  
-  Field2D dx, dy; ///< Mesh spacing in x and y
-  BoutReal dz; ///< Mesh spacing in Z
 
-  BoutReal zlength() const { return dz * nz; } ///< Length of the Z domain. Used for FFTs
+  metric_field_type dx, dy, dz; ///< Mesh spacing in x and y
+  // BoutReal dz; ///< Mesh spacing in Z
+
+  Field2D zlength() const {
+#ifdef COORDINATES_USE_3D
+    Field2D result(0., localmesh);
+    BOUT_FOR_SERIAL(i, dz.getRegion("RGN_ALL")) { result[i] += dz[i]; }
```

Comment:
Not so sure about this, `result` is 2D and `dz` is 3D here so the loop indices are a little deceiving as we're summing up all the z-points.
Presumably we need a reduction to use OpenMP, but it needs to be over just the z-points, so something like:

```cpp
    BOUT_OMP(parallel for reduction(+:result[i])
    BOUT_FOR_SERIAL(i, dz.getRegion("RGN_ALL")) { result[i] += dz[i]; }
```

except I don't think that will work!

Reply:
Ah, I see! My mistake, sorry.

I think this method does get called in places that will run every rhs call though - I think it might be worth having a `zlength` member variable rather than a method so we can do the calculation just once in the constructor.

Reply:
The method is nice because we can keep it read-only, but yes it might be nice to cache the result. But then it immediately requires more of a refactor of `Coordinates` to have getters/setters for all the members -- changing `dz` will change `zlength`.

I do think that's worth doing, but in another PR.

I'd argue for keeping this as-is for now. Might not be the most efficient, but we need other things in place first

Reply:
No problem leaving as is for now.
I think we could treat zlength like the metric components, etc. It seems unlikely that someone will change `dz` in the first place, but if they do then we could update `zlength` in `Coordinates::geometry()`. The "call `Coordinates::geometry()` if you change something" approach isn't the safest - but is it more efficient, because it avoids going through getters inside tight loops: `coords->g_22[i]` vs. `coords->g_22()[i]`? It's true we could move all the calls to getters outside the loops and keep the fields in local variables, but that might require a lot of refactoring!